### ISSUE TITLE
Switch leaflet for openlayers?

### ISSUE THREAD

**Issue by nilsnolde** [gis-ops/valhalla-app#158]
### What would you like to share?

WDYT @chrstnbwnkl ?

I asked Stadia Maps if we can use their Alidade light style for background maps. While they of course have raster data as well, I'd prefer to then consume their MVT endpoint, it's just so much nicer for end users.. 

Apart from the fact (not really fact, but looks very much like it) that leaflet is at the very least carrying tech debt that's hard to overcome without a major re-write and can't see that happening anytime soon. 

Uff, let's see, this seems to become a serious re-write of our app with the class -> functional React changes. It could be a bit messy to do it all at the same time, maybe first the full React re-write, then this?! We'll see, I just wanted to put it out there:)

**Comment by chrstnbwnkl** [comment]
Agree! Definitely a big work package, but probably worth the effort. We can just leave it here now and discuss it when GSoC starts :-)

**Comment by nilsnolde** [comment]
That's was "google summer of code". That year we wanted to slip in this project but ended up not to. 

I also saw recently that leaflet got some new traction and they had a big release. Didn't look if that finally included supporting vector tiles. Still, I'd like to switch tbh, but personally don't have enough clue to have an opinion which one of the 2 alternatives (openlayers or maplibre) is better for our particular project.

**Comment by nilsnolde** [comment]
Urgh yeah, that seems like there's smth blocking the UI. Thanks for the PoC, it's very apparent to me too. I can't ever recall hearing about anything like that, @chrstnbwnkl used it a lot in the past, hope he can spot a problem with the openlayers usage.

**Comment by nilsnolde** [comment]
I asked chatgpt, just for fun, but it did come up with a few good suggestions. First good info is that maplibre renders in service workers using the GPU so main thread stays unaffected by zooming, maplibre apparently doesn't. However, there seem to be this: `Port your style to OL’s OffscreenCanvas worker setup or OL’s WebGL vector tile layer. That gets you closest to MapLibre’s threading/GPU pipeline within OL’s ecosystem.`

It also gave some hints where performance can be optimized as well. Some make immediate sense like `updateWhileInteracting: false,`, others I can't really assess without more research (e.g. `cacheSize`, should be high probably?), others are pointless probably (e.g. `maxZoom`, OSM only publishes up to z14):
```js
import Map from 'ol/Map';
import View from 'ol/View';
import MVT from 'ol/format/MVT';
import VectorTileLayer from 'ol/layer/VectorTile';
import VectorTileSource from 'ol/source/VectorTile';

// Vector-tile source: basic PBF loader
const vtSource = new VectorTileSource({
  format: new MVT(),
  url: 'https://…/{z}/{x}/{y}.pbf',

  maxZoom: 14,       // don’t request super-fine tiles you’ll never display
  cacheSize: 512,    // larger in-memory tile cache = fewer re-parses
  overlaps: false,   // skip expensive geometry checks if tiles are perfectly cut
  tilePixelRatio: 1, // avoids drawing extra pixels for HiDPI during interaction
});

// Vector-tile layer with performance-oriented flags
const vtLayer = new VectorTileLayer({
  source: vtSource,

  renderMode: 'hybrid',  // scales tiles as images while zooming, keeps vectors crisp when still
  declutter: false,      // skip collision-detection/layout during continuous zoom
  updateWhileInteracting: false, // don’t re-style on every wheel/touch move
  updateWhileAnimating: false,   // same for animated zooms/pans

  // style: preCreatedStyleOrCachedFunction
  //   → supply a cached or pre-instantiated Style to avoid GC thrash
});

// Map container
const map = new Map({
  target: 'map',

  pixelRatio: 1, // force 1× device pixel ratio to cut canvas work (test impact)
  layers: [vtLayer],

  view: new View({
    center: [0, 0],
    zoom: 3,
  }),
**});**
```

**Comment by mustaphaturhan** [comment]
@nilsnolde I also tried with claude-4-sonnet but I think the problem is, the `style.json` file I got from `openstreetmap` huge to handle for openlayers because I tried more lightweight version from `https://tiles.versatiles.org/assets/styles/colorful/style.json` and it worked better. ([click to see an example](https://openlayers.org/ol-mapbox-style/examples/versatiles.html)). On the other hand, I am not able to test `VectorTileSource` because I have a `style.json` file written as [maplibre style object](https://maplibre.org/maplibre-style-spec/) and `openlayers` is not compatible with it without `ol-mapbox-style` [plugin](https://openlayers.org/ol-mapbox-style/).

**Comment by chrstnbwnkl** [comment]
I haven't used OpenLayers with huge vector tiles and complex stylesheets yet, so I can't really comment on it, but I agree that MapLibre is by far the smoothest in this example, even comparing against the lightweight version. Maybe MapLibre really is better, if we want to use those shiny new OSM vector tiles. 

For geometry drawing I saw that https://github.com/JamesLMilner/terra-draw has gained quite a bit of popularity. Never used it myself, but seems like the better alternative to https://github.com/mapbox/mapbox-gl-draw when using MapLibre, since the latter has not really been maintained for ages...

**Comment by nilsnolde** [comment]
Yep that's fine with us @mustaphaturhan, thanks!
